### PR TITLE
Remove dependency on System.Console in Blazor WASM app

### PR DESCRIPTION
**Pull request by eerhardt**
This is the first change necessary. It removes the one place that depends on `Console.WriteLine`.

This first change doesn't save much size (.br compressed):

* **Before**: 2,719,214 bytes
* **After**: 2,719,111 bytes

However, if we can also remove the last place that references `System.Console`, we should be able to trim the whole assembly (6KB .br compressed).

https://github.com/dotnet/aspnetcore/blob/dfe625fbd122557cc4dda0f3d6f3dc5c7afc8c8e/src/Components/WebAssembly/WebAssembly/src/Hosting/EntrypointInvoker.cs#L79-L83

Speaking with @lewing, we thought if we unified the startup paths between the mono runtime and Blazor, we should be able to remove this last dependency on `System.Console`. @lewing was going to log an issue to track unifying the startup code.

cc @marek-safar

### CODE REVIEW

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (Eilon) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: The parameter to ArgumentException's ctor is a message, not a parameter name:

https://docs.microsoft.com/en-us/dotnet/api/system.argumentexception.-ctor?view=net-5.0#System_ArgumentException__ctor_System_String_

There is an overload that takes both, though (but get the order right 😄 ):

https://docs.microsoft.com/en-us/dotnet/api/system.argumentexception.-ctor?view=net-5.0#System_ArgumentException__ctor_System_String_System_String_

Comment (Eilon) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: Also, this is now throwing an exception where it didn't used to, so perhaps a breaking change (in a super obscure case)? Maybe just call jsRuntime w/ console.error(...) as seen in earlier lines?

Comment (eerhardt) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: > The parameter to ArgumentException's ctor is a message, not a parameter name:

(doh) I swear I've been bit by this too many times.... Thanks for the catch.

> Also, this is now throwing an exception where it didn't used to, so perhaps a breaking change

It's for the case where someone passes in an invalid `LogLevel` enum into `Log`. Is changing this to an exception not acceptable? I would think devs who are casting random ints to `LogLevel` would want to know they have a bug in their code.

We throw for other loggers:

https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs#L101
https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/Microsoft.Extensions.Logging.Console/src/SystemdConsoleFormatter.cs#L115
https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/libraries/Microsoft.Extensions.Logging.Console/src/SimpleConsoleFormatter.cs#L152

Comment (Eilon) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: No idea whether it's acceptable, but it's a change. I agree it's super obscure. I'll let one of the code owners help answer that as I am not a stakeholder.

Comment (eerhardt) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: @pranavkm @SteveSandersonMS - any thoughts here? Is changing this to an exception acceptable?

Comment (Eilon) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: (BTW the ArgNull parameter value still needs to be fixed.)

Comment (eerhardt) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: Agreed - but I only want to fix it once. If we can't even throw an exception here, I won't bother an intermediate change.

Comment (pranavkm) [reply src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: @captainsafia are you able to look at this change? IMO, throwing for an unsupported log level does not seem unreasonable.

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (Eilon) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]: Odd grammar here

**Review by Eilon — APPROVED** [review]
I approve as a non-stakeholder.

**Review by javiercn — APPROVED** [review]

**Comment by javiercn** [comment]
> However, if we can also remove the last place that references `System.Console`, we should be able to trim the whole assembly

I think it could be doable, we would need to call some JS interop API instead to notify JS of the error. @SteveSandersonMS do you have any thoughts/concerns?

**Comment by eerhardt** [comment]
@captainsafia - any thoughts on this PR? I think it is ready to merge, unless you have other feedback.

**Comment by javiercn** [comment]
@eerhardt can we let @SteveSandersonMS comment on the last System.Console usage to see if we can just update this PR and trim the entire assembly?

**Comment by eerhardt** [comment]
> if we can just update this PR and trim the entire assembly?

My understanding is to trim the entire System.Console assembly is going to take a bit of refactoring how we initialize a Blazor WASM app, and it may take coordination with @lewing's team.

**Comment by javiercn** [comment]
> My understanding is to trim the entire System.Console assembly is going to take a bit of refactoring how we initialize a Blazor WASM app, and it may take coordination with @lewing's team.

Ah, thanks for the clarification.

Then let's get this merged!

Thanks for the PR

**Comment by javiercn** [comment]
@eerhardt do you want to merge the PR yourself or are you ok if we do it for you

File: src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs
Comment (captainsafia) [inline src/Components/WebAssembly/WebAssembly/src/Services/WebAssemblyConsoleLogger.cs:105]:
```suggestion
                            _jsRuntime.InvokeVoid("console.log", formattedMessage);
```

If we can avoid a breaking change here, it would be great. The above would be an adequate substitute.

**Comment by SteveSandersonMS** [comment]
> Speaking with @lewing, we thought if we unified the startup paths between the mono runtime and Blazor, we should be able to remove this last dependency on System.Console. @lewing was going to log an issue to track unifying the startup code.

I'm sure we can remove the `Console.Error.WriteLine` call from `HandleStartupException` independently of any unification to the startup paths. Since that call is coming from the `.WebAssembly` assembly, it could directly call `WebAssembly.JSInterop.InternalCalls.InvokeJS` to get the log message into JS world.

Or if that's inconvenient, we could even have a dedicated `InternalCall` method specifically for logging errors, which would bypass all dependencies on any JS interop  (and maybe there already is one).

**Review by captainsafia — APPROVED** [review]
Looks good from my end!

We can take Steve's recommendation on cleaning up the `System.Console` use in the entrypoint invoker and finish this off.

**Pull request merged**
